### ISSUE TITLE
Pre-trained model link is invalid

### ISSUE THREAD

**Issue by geert56** [aliostad/deep-learning-lang-detection#1]
The link to the pre-trained model is no longer working. The file has vanished on Google drive.

How many epochs was the model trained for?
I try to repeat the experiment using only the "train" files of this github and the default
20 epochs of training. I cannot get the results reported in the README of 99% acc.

Thanks.

**Comment by aliostad** [comment]
What do you get? You should be very close to 97-98%? 99% was the best run.

Yeah I removed the file since I had issues with my google drive. I can supply that back.

**Comment by geert56** [comment]
Dear Ali,
Just cloning your github and running "train.py" for the default 20 epochsI get close to 60% validation accurcacy.
Here are the charts:

    On Friday, April 23, 2021, 3:42:17 AM EDT, Ali Kheyrollahi ***@***.***> wrote:  
 
 


What do you get? You should be very close to 97-98%? 99% was the best run.

**Comment by geert56** [comment]
Dear Ali,
If I train for 100 epochs I get decent results.See the charts and this test_run.sh output:
Final result: 13103/16016 (0.8181193806193806)
go - Precision:0.9896073903002309 Recall: 0.8561438561438561
csharp - Precision:0.9920634920634921 Recall: 0.8741258741258742
java - Precision:0.9683026584867076 Recall: 0.9460539460539461
js - Precision:0.9885807504078303 Recall: 0.6053946053946054
cpp - Precision:0.9809885931558935 Recall: 0.7732267732267732
ruby - Precision:0.9969742813918305 Recall: 0.6583416583416584
powershell - Precision:0.974390243902439 Recall: 0.7982017982017982
bash - Precision:0.9844097995545658 Recall: 0.8831168831168831
php - Precision:0.998911860718172 Recall: 0.9170829170829171
css - Precision:0.9873684210526316 Recall: 0.9370629370629371
xml - Precision:0.9867075664621677 Recall: 0.964035964035964
python - Precision:0.9844236760124611 Recall: 0.6313686313686314
scala - Precision:0.9790897908979089 Recall: 0.7952047952047953
clojure - Precision:0.9892818863879957 Recall: 0.922077922077922
fsharp - Precision:0.9742990654205608 Recall: 0.8331668331668332
sql - Precision:0.96398891966759 Recall: 0.6953046953046953
go:             857/1001 (0.8561438561438561)
csharp:         875/1001 (0.8741258741258742)
java:           947/1001 (0.9460539460539461)
js:             606/1001 (0.6053946053946054)
cpp:            774/1001 (0.7732267732267732)
ruby:           659/1001 (0.6583416583416584)
powershell:             799/1001 (0.7982017982017982)
bash:           884/1001 (0.8831168831168831)
php:            918/1001 (0.9170829170829171)
css:            938/1001 (0.9370629370629371)
xml:            965/1001 (0.964035964035964)
python:         632/1001 (0.6313686313686314)
scala:          796/1001 (0.7952047952047953)
clojure:                923/1001 (0.922077922077922)
fsharp:         834/1001 (0.8331668331668332)
sql:            696/1001 (0.6953046953046953)
Sum of Squared Error: 2546.909804087411

I converted your code to Python 3 and am running using TF 2.5.0and its built-in Keras. I use a single V100 GPU; runtime is an hour or so.


    On Friday, April 23, 2021, 3:42:17 AM EDT, Ali Kheyrollahi ***@***.***> wrote:  
 
 


What do you get? You should be very close to 97-98%? 99% was the best run.

**Comment by geert56** [comment]
Oops, that was a Tesla P100 not V100.

    On Friday, April 23, 2021, 3:42:17 AM EDT, Ali Kheyrollahi ***@***.***> wrote:  
 
 


What do you get? You should be very close to 97-98%? 99% was the best run.

**Comment by aliostad** [comment]
Well, I think I had access to a powerful GPU box when I was training but now I don't. I would easily get 95-98% back then.
Let me try it again.

**Comment by aliostad** [comment]
Doesn't run on my box/

**Comment by aliostad** [comment]
OK, I uploaded the model: https://drive.google.com/file/d/1vKnPe2PltKetLblj4BzWiFD4RDqaeoD_/view?usp=sharing

I get decent results from running bunch of snippets.
Also bear in mind there are errors in the data so some labels are incorrect and it cannot achieve perfect result as such.

Just drop `save_tmp.h5` file in the project root and run `lang_detect_api.py`. If you get error, change the last line by adding `threaded=False`:

``` python
app.run(debug=False, threaded=False, use_reloader=False, host='0.0.0.0')
```

**Comment by geert56** [comment]
Thanks Ali, for the model.

I see that yours is smaller than mine:
-rw-rw-r-- 1 geert geert 202773808 Apr 23 12:32 model.h5
Are you sure that the github is up-to-date?You did not change any parameters?

    On Saturday, April 24, 2021, 3:05:54 PM EDT, Ali Kheyrollahi ***@***.***> wrote:  
 
 


OK, I uploaded the model: https://drive.google.com/file/d/1vKnPe2PltKetLblj4BzWiFD4RDqaeoD_/view?usp=sharing

I get decent results from running bunch of snippets.
Also bear in mind there are errors in the data so some labels are incorrect and it cannot achieve perfect result as such.

Just drop save_tmp.h5 file in the project root and run lang_detect_api.py. If you get error, change the last line by adding threaded=False:
app.run(debug=False, threaded=False, use_reloader=False, host='0.0.0.0')

**Comment by geert56** [comment]
Here are the results I get running test_run.sh on your model save_tmp.h5:
Final result: 15877/16016 (0.9913211788211789)
    go - Precision:0.9969909729187563 Recall: 0.993006993006993
    csharp - Precision:0.9900793650793651 Recall: 0.997002997002997
    java - Precision:0.9910891089108911 Recall: 1.0
    js - Precision:0.9899396378269618 Recall: 0.983016983016983
    cpp - Precision:0.9899799599198397 Recall: 0.987012987012987
    ruby - Precision:0.9939698492462311 Recall: 0.988011988011988
    powershell - Precision:0.9940179461615155 Recall: 0.996003996003996
    bash - Precision:0.9929789368104313 Recall: 0.989010989010989
    php - Precision:0.9979939819458375 Recall: 0.994005994005994
    css - Precision:0.9959758551307847 Recall: 0.989010989010989
    xml - Precision:0.994005994005994 Recall: 0.994005994005994
    python - Precision:0.9764243614931237 Recall: 0.993006993006993
    scala - Precision:0.991008991008991 Recall: 0.991008991008991
    clojure - Precision:0.995 Recall: 0.994005994005994
    fsharp - Precision:0.9880834160873883 Recall: 0.994005994005994
    sql - Precision:0.9969481180061037 Recall: 0.9790209790209791
    go:             994/1001 (0.993006993006993)
    csharp:         998/1001 (0.997002997002997)
    java:           1001/1001 (1.0)
    js:             984/1001 (0.983016983016983)
    cpp:            988/1001 (0.987012987012987)
    ruby:           989/1001 (0.988011988011988)
    powershell:             997/1001 (0.996003996003996)
    bash:           990/1001 (0.989010989010989)
    php:            995/1001 (0.994005994005994)
    css:            990/1001 (0.989010989010989)
    xml:            995/1001 (0.994005994005994)
    python:         994/1001 (0.993006993006993)
    scala:          992/1001 (0.991008991008991)
    clojure:                995/1001 (0.994005994005994)
    fsharp:         995/1001 (0.994005994005994)
    sql:            980/1001 (0.9790209790209791)
    Sum of Squared Error: 221.68742297034717



    On Saturday, April 24, 2021, 3:05:54 PM EDT, Ali Kheyrollahi ***@***.***> wrote:  
 
 


OK, I uploaded the model: https://drive.google.com/file/d/1vKnPe2PltKetLblj4BzWiFD4RDqaeoD_/view?usp=sharing

I get decent results from running bunch of snippets.
Also bear in mind there are errors in the data so some labels are incorrect and it cannot achieve perfect result as such.

Just drop save_tmp.h5 file in the project root and run lang_detect_api.py. If you get error, change the last line by adding threaded=False:
app.run(debug=False, threaded=False, use_reloader=False, host='0.0.0.0')

**Comment by aliostad** [comment]
So are you happy with this result? I think I might have left it running for many epochs hence it got to a good place. I myself am sometimes surprised how good it does it although some languages are very simple (such as clojure where it starts with an opening bracket).

The reason I started training on snippets too I saw it was cheating and was learning only the import statements at the top of the files. But I could show that it can indeed learn the language patterns themselves.

Let me know if I can close this issue.

**Comment by geert56** [comment]
I am happy with the results I get from your "save_tmp.h5" pretrained model.Could you please update the github. You still have not told me for how manyepochs you trained; for sure not the 20 that's in the code.I also wonder why your h5 model is smaller than the one you get by trainingdirectly using "train.py" from the github. So again, please verify that yourlocal copy agrees with the github (git push).
Thanks for the conversation. I guess you can now close the issue.

    On Sunday, April 25, 2021, 4:05:06 PM EDT, Ali Kheyrollahi ***@***.***> wrote:  
 
 


So are you happy with this result? I think I might have left it running for many epochs hence it got to a good place. I myself am sometimes surprised how good it does it although some languages are very simple (such as clojure where it starts with an opening bracket).

The reason I started training on snippets too I saw it was cheating and was learning only the import statements at the top of the files. But I could show that it can indeed learn the language patterns themselves.

Let me know if I can close this issue.

**Comment by aliostad** [comment]
Hi,

Sorry but I cannot remember! This repo is actually at least 4 year old although you might not see it. 
I had some test files that I had harvested from github but they contained some sensitive information (which not really my fault, they were public when I harvested) and then they realised and removed but the copy had stayed in my repo.

So short answer is I cannot remember but I might have tinkered with the model hence now it might look different. You can always see the summary of the network and compare to the code.

I will update the link.

**Comment by achmedzhanov** [comment]
> I am happy with the results I get from your "save_tmp.h5" pretrained model.Could you please update the github. You still have not told me for how manyepochs you trained; for sure not the 20 that's in the code.I also wonder why your h5 model is smaller than the one you get by trainingdirectly using "train.py" from the github. So again, please verify that yourlocal copy agrees with the github (git push). 

such a high accuracy of the model may indicate that the model was trained on the data on which it is tested :)
